apps/mark-scan/accessible-controller/src/port.rs

```diff
@@ -18,15 +17,52 @@ pub struct Port {
 }
 
 impl Port {
-    pub fn open() -> color_eyre::Result<Self> {
-        let inner = serialport::new(DEVICE_PATH, DEVICE_BAUD_RATE)
+    pub fn open_by_ids(vendor_id: u16, product_id: u16) -> color_eyre::Result<Self> {
+        let device_path = Self::get_device_path(vendor_id, product_id)?;
+        Self::open(&device_path)
+    }
+
+    fn open(path: &str) -> color_eyre::Result<Self> {
+        let inner = serialport::new(path, DEVICE_BAUD_RATE)
             .timeout(POLL_INTERVAL)
             .open()?;
         let mut port = Self { inner };
         port.validate_connection()?;
         Ok(port)
     }
 
+    fn get_device_path(vendor_id: u16, product_id: u16) -> Result<String, io::Error> {
+        match available_ports() {
+            Ok(ports_info) => {
+                for port_info in ports_info {
```

Comment:
Wasn't sure if `iter` is more idiomatic but this is how `serialport` examples do it

Reply:
No, this is the typical way to do it in Rust. `for` loops desugar into a `loop` with manual handling of an `Iterator`:

```rust
// this…
for i in 0..2 {
    println!("{}", i);
}

// becomes (more or less) this:
{
    let _t =
        match IntoIterator::into_iter(Range { start: 0, end: 2 }) {
            mut iter =>
            loop {
                match iter.next() {
                    None => break,
                    Some(i) => {
                        println!("{}", i);
                    },
                }
            },
        };
    _t
}
```

`IntoIterator::into_iter` consumes `self`, which in your code is `ports_info`. This is what you want since you don't need to use `ports_info` later. If you did, you could do `for port_into in ports_info.iter()`, which creates an `Iterator<Item = &SerialPortInfo>` instead of `Iterator<Item = SerialPortInfo>`, making `port_info` a reference instead of an owned value.

Reply:
Thank you for the clarification!